src/engraving/libmscore/note.cpp

```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
 
 void Note::setDotY(DirectionV pos)
 {
-    bool onLine = false;
     double y = 0;
-
+    bool onLine = !(line() & 1);
     if (staff()->isTabStaff(chord()->tick())) {
         // with TAB's, dotPosX is not set:
         // get dot X from width of fret text and use TAB default spacing
         const Staff* st = staff();
         const StaffType* tab = st->staffTypeForElement(this);
         if (tab->stemThrough()) {
             // if fret mark on lines, use standard processing
-            if (tab->onLines()) {
-                onLine = true;
-            } else {
+            if (!tab->onLines()) {
                 // if fret marks above lines, raise the dots by half line distance
                 y = -0.5;
             }
+            if (pos == DirectionV::AUTO) {
+                bool oddVoice = voice() & 1;
+                y = oddVoice ? 0.5 : -0.5;
+            } else if (pos == DirectionV::UP) {
+                y = -0.5;
+            } else {
+                y = 0.5;
+            }
         }
         // if stems beside staff, do nothing
         else {
             return;
         }
-    } else {
-        onLine = !(line() & 1);
-    }
-
-    bool oddVoice = voice() & 1;
-    if (onLine) {
-        // displace dots by half spatium up or down according to voice
-        if (pos == DirectionV::AUTO) {
-            y = oddVoice ? 0.5 : -0.5;
-        } else if (pos == DirectionV::UP) {
-            y = -0.5;
+    } else if (onLine) {
+        // NON-TAB
+        bool oddVoice = voice() & 1;
+        Chord* c = chord();
+        bool hasVoices = c->measure()->hasVoices(c->staffIdx(), c->tick(), c->ticks());
+        std::vector<Note*> topDownNotes;
+        std::vector<Note*> bottomUpNotes;
+        std::vector<size_t> anchoredDots;
+        // construct combined chords using the notes from overlapping chords
+        if (!hasVoices) {
+            // only this voice, so topDownNotes is just the notes in the chord
+            for (Note* note : c->notes()) {
+                if (note->line() & 1) {
+                    int newOffset = 0;
+                    bool adjustDown = (c->voice() & 1) && !c->up();
+                    if (!anchoredDots.empty() && anchoredDots.back() == note->line()) {
+                        if (anchoredDots.size() >= 1 && anchoredDots[anchoredDots.size() - 2] == note->line() + adjustDown ? 2 : -2) {
+                            newOffset = adjustDown ? -2 : 2;
```

Comment:
What does these numbers (-2 and 2) mean?

Reply:
Number of staff spaces or lines to adjust, where the top line is 0, then the top space is 1, etc etc